pkgs/ffigen/lib/src/code_generator/objc_methods.dart

```diff
@@ -0,0 +1,224 @@
+    }
+  }
+
+  ObjCMethod _maybeReplaceMethod(ObjCMethod? oldMethod, ObjCMethod newMethod) {
+    if (oldMethod == null) return newMethod;
+
+    // Typically we ignore duplicate methods. However, property setters and
+    // getters are duplicated in the AST. One copy is marked with
+    // ObjCMethodKind.propertyGetter/Setter. The other copy is missing
+    // important information, and is a plain old instanceMethod. So if the
+    // existing method is an instanceMethod, and the new one is a property,
+    // override it.
+    if (newMethod.isProperty && !oldMethod.isProperty) {
+      return newMethod;
+    } else if (!newMethod.isProperty && oldMethod.isProperty) {
+      // Don't override, but also skip the same method check below.
+      return oldMethod;
+    }
+
+    // Check the duplicate is the same method.
+    if (!newMethod.sameAs(oldMethod)) {
+      _logger.severe('Duplicate methods with different signatures: '
+          '$originalName.${newMethod.originalName}');
+      return newMethod;
+    }
+
+    // There's a bug in some Apple APIs where an init method that should return
+    // instancetype has a duplicate definition that instead returns id. In that
+    // case, use the one that returns instancetype. Note that since instancetype
+    // is an alias of id, the sameAs check above passes.
+    if (ObjCBuiltInFunctions.isInstanceType(newMethod.returnType) &&
+        !ObjCBuiltInFunctions.isInstanceType(oldMethod.returnType)) {
+      return newMethod;
+    } else if (!ObjCBuiltInFunctions.isInstanceType(newMethod.returnType) &&
+        ObjCBuiltInFunctions.isInstanceType(oldMethod.returnType)) {
+      return oldMethod;
+    }
+
+    return newMethod;
+  }
+
+  bool _shouldIncludeMethod(ObjCMethod method) =>
+      method.childTypes.every((Type t) {
+        t = t.typealiasType.baseType;
+
+        // Ignore methods with variadic args.
+        // TODO(https://github.com/dart-lang/native/issues/1192): Remove this.
+        if (t is Struct && t.originalName == '__va_list_tag') {
+          return false;
+        }
+
+        // Ignore methods with block args or rets when we're generating in
+        // package:objective_c.
+        // TODO(https://github.com/dart-lang/native/issues/1180): Remove this.
+        if (builtInFunctions.generateForPackageObjectiveC && t is ObjCBlock) {
+          return false;
+        }
+
+        return true;
+      });
+}
+
+enum ObjCMethodKind {
+  method,
+  propertyGetter,
+  propertySetter,
+}
+
+class ObjCProperty {
+  final String originalName;
+  String? dartName;
+
+  ObjCProperty(this.originalName);
+}
+
+class ObjCMethod {
+  final String? dartDoc;
+  final String originalName;
+  final ObjCProperty? property;
+  Type returnType;
+  final List<ObjCMethodParam> params;
+  final ObjCMethodKind kind;
+  final bool isClass;
+  final bool isOptional;
+  bool returnsRetained = false;
+  ObjCInternalGlobal? selObject;
+  ObjCMsgSendFunc? msgSend;
+  ObjCBlock? protocolBlock;
```

Comment:
Should this be `late final`? E.g. should be assigned once (and be explicitly assigned null for if it's supposed to be null), and throw if accessed before it's assigned instead of returning null.

Reply:
Done